analysis_test.go

```diff
@@ -1,62 +1,57 @@
 package main
 
 import (
-	"bytes"
 	"errors"
-	"io"
-	"io/ioutil"
 	"testing"
 )
 
-var b *bytes.Buffer
-
-func mockOutFor(project, resource string) (io.Writer, io.Closer, error) {
-	return b, ioutil.NopCloser(nil), nil
-}
-
 func mockCheckFactoryOnePassOneFail() []CheckTask {
 	return []CheckTask{mockTestOne, mockTestTwo}
 }
 func mockCheckFactoryOnePass() []CheckTask {
 	return []CheckTask{mockTestOne}
 }
 
-func mockTestOne(project string, stdErr io.Writer) (Result, error) {
+func mockTestOne(project, basepath string) (Result, error) {
 	result := Result{StatusMessage: "Called mockTestOne"}
 	return result, nil
-
 }
 
-func mockTestTwo(project string, stdErr io.Writer) (Result, error) {
+func mockTestTwo(project, basepath string) (Result, error) {
 	result := Result{StatusMessage: "Called mockTestTwo"}
 	return result, errors.New("FAIL")
 }
 
-func TestCheckTasks(t *testing.T) {
-	b = bytes.NewBuffer([]byte{})
-	task := checkTasks(mockCheckFactoryOnePassOneFail, "MockProject", mockOutFor, mockOutFor)
+func TestCheckTasksWithFail(t *testing.T) {
+	results := make(chan CheckResults, 1)
+	defer close(results)
+	task := checkProjectTask(mockCheckFactoryOnePassOneFail, "MockProject", "/", results)
+
 	err := task()
 
 	if err == nil {
 		t.Fatal("Expected error")
 	}
 
-	if len(string(b.Bytes())) == 0 {
-		t.Fatal("Tests not called")
-
+	result := <-results
+	if (len(result.Results)) != 2 {
```

Comment:
> The task runs several checks inside that project synchronously.

Does it need to be synchronous?

---

It might be worth exploring that idea. In my head I was trying to get to task == check, so that we have a simple execution model: concurrently put tasks to a "bag" and execute them.

For things that we want to execute per project, there would be as many of those tasks as projects (so that they can be executed concurrently).

I see the original Task signature, `func() error` (is it?), doesn't help us grouping results as we want in the "analysis tasks"....